src/store.rs

```diff
@@ -25,11 +28,40 @@ pub fn get_missing_files(root: &str, wanted: &[String]) -> io::Result<Vec<String
             Ok(true) => None,
             Err(e) => Some(Err(e)),
         })
-    .collect()
+        .collect()
+}
+
+pub async fn put_file(root: &str, mut file: TempFile<'_>, hash: &str) -> io::Result<String> {
+    let temp_path = std::env::temp_dir().join(hash);
+    file.persist_to(&temp_path).await?;
+    let alg = config::read_config(root)?.core.hash_algorithm;
+    let content = fs::read_to_string(&temp_path)?;
+    let valid_hash = hash::hash_data(content, alg);
+    if hash != valid_hash {
+        return Err(io::Error::new(ErrorKind::InvalidInput,
+                                  format!("Hash {} does not match file contents. Expected {}",
+                                          hash, valid_hash)));
+    }
+
+    let path = file_path(root, hash)?;
+    let pathname = (path).to_str()
+        .map(String::from)
+        .unwrap();
+
+    if !file_exists(root, hash)? {
+        fs::create_dir_all(path.parent().unwrap())?;
+        fs::rename(temp_path, path)
```

Comment:
This is potentially problematic (either causing a failure or a copy and associated performance cost), depending on how temp_dir() is resolved because it _must_ be on the same file system for rename to work as expected. I believe that in outpack (or outpack.server) for this we used a temporary directory within the .outpack tree to enforce this.

This is a non-theoretical issue on at least some configurations of macOS where the temporary directory is at some user-specific mount which is a different logical file system and an atomic rename does not work.

Reply:
So an issue for the case where this is used on metal as a cli? Ok, have changed to create a temp dir inside the outpack root...